api/rtx/src/uvisor_allocator.c

```diff
@@ -0,0 +1,215 @@
+#include <stdlib.h>
+#include "uvisor-lib/uvisor-lib.h"
+#include "rt_TypeDef.h"
+#include "rt_Memory.h"
+#include <string.h>
+
+#define DPRINTF(...) {};
+/* #define DPRINTF(fmt, ...) printf(fmt, ## __VA_ARGS__) */
+
+/* Declare this variable here, so the tier-2 allocator _always_ uses the
+ * page size that the tier-1 allocator expects! */
+const uint32_t __uvisor_page_size = UVISOR_PAGE_SIZE;
+
+/* Internal structure currently only contains the page table */
+typedef struct {
+    UvisorPageTable table;
+} UvisorAllocatorInternal;
+
+static inline UvisorPageTable * table(UvisorAllocator allocator) {
+    return &(((UvisorAllocatorInternal *) allocator)->table);
+}
+
+UvisorAllocator uvisor_allocator_create_with_pool(
+    void * mem,
+    size_t bytes)
+{
+    UvisorAllocatorInternal * allocator = mem;
+    /* Signal that this is non-page allocated memory */
+    allocator->table.page_size = bytes;
+    allocator->table.page_count = 0;
+    /* The MEMP structure must be placed AFTER table.page_origins[0] !!! */
+    size_t offset = offsetof(UvisorAllocatorInternal, table.page_origins) + sizeof(void *);
+    /* Create MEMP structure inside the memory */
+    if (rt_init_mem(mem + offset, bytes - offset)) {
+        /* Abort if failed */
+        DPRINTF("uvisor_allocator_create_with_pool: MEMP allocator creation failed\n\n");
+        return NULL;
+    }
+    /* Remember the MEMP pointer though */
+    allocator->table.page_origins[0] = mem + offset;
+    DPRINTF("uvisor_allocator_create_with_pool: Created MEMP allocator %p with offset %d\n\n", mem + offset, offset);
+    return allocator;
+}
+
+UvisorAllocator uvisor_allocator_create_with_pages(
+    size_t size,
+    size_t maximum_malloc_size)
+{
+    /* Additional overhead for each block */
+    const size_t block_overhead = 2 * sizeof(MEMP);
+    size_t page_count = 0;
+
+    /* Calcutate the required total size considering page size and overhead */
+    for (int request = size; request > 0; request -= UVISOR_PAGE_SIZE, page_count++)
+    {
+        /* Adjust the requested and total sizes */
+        request += block_overhead;
+        size += block_overhead;
+    }
+    DPRINTF("uvisor_allocator_create_with_pages: Requesting %u pages for at least %uB\n", page_count, size);
+
+    /* Compute the maximum allocation within our blocks */
+    size_t maximum_allocation_size = UVISOR_PAGE_SIZE - block_overhead;
+    /* If the required maximum allocation is larger than we can provide, abort */
+    if (maximum_malloc_size > maximum_allocation_size) {
+        DPRINTF("uvisor_allocator_create_with_pages: Maximum allocation request %uB is larger then available %uB\n\n", maximum_malloc_size, maximum_allocation_size);
+        return NULL;
+    }
+
+    /* Compute the require memory size for the page table */
+    size_t allocator_type_size = sizeof(UvisorAllocatorInternal);
+    /* Add size for each additional page */
+    allocator_type_size += (page_count - 1) * sizeof(void *);
```

Comment:
The `void *` refers to the type of [the `void * page_origins` here](https://github.com/ARMmbed/uvisor/blob/dev/api/inc/page_allocator_exports.h#L58).